### PR TITLE
Periodic task for processing consents that are expired

### PR DESCRIPTION
**Pull request by ajeety4**
## Product Description
The ABDM Gateway was supposed to send notification via API call once the consent is expired. However this has been extremely unreliable and after latest discussion on ABDM forum [here](https://devforum.abdm.gov.in/t/not-getting-hiu-notify-for-expired-consent/6853/8?u=ayadav), it seems they would not send notifications for consent expiry.
This PR adds the feature for processing consents that are expired using celery periodic tasks. This will not cause any side effect in case ABDM still sends notification for consent expiry.

## Technical Summary

Includes below implementation:
- Dynamically derive status field for HIU consent requests
- Wrapper to create periodic task as decorator similar to as in HQ
- HIP periodic task to handle expired consent and artefacts
- HIU periodic task to handle expired artefacts.

[Internal Specifications Doc](https://docs.google.com/document/d/1gbQcAC3H4kj22iMCMMKSYMVO5_7kt56qiS8OWkd6e34/edit#heading=h.f9dgxbqmorgx)

## Automated test coverage
Test Cases added for HIU. Plan to add test cases for HIP consents.

### CODE REVIEW

File: abdm_integrator/hip/tasks.py [hunk 1]
```diff
@@ -32,3 +37,12 @@ def process_patient_care_context_link_init_request(self, request_data):
 def process_patient_care_context_link_confirm_request(self, request_data):
     from abdm_integrator.hip.views.care_contexts import GatewayCareContextsLinkConfirmProcessor
     GatewayCareContextsLinkConfirmProcessor(request_data).process_request()
+
+
+@CELERY_PERIODIC_TASK(run_every=crontab(hour='*/2', minute='0'), queue=app_settings.CELERY_QUEUE)
+def process_hip_expired_consents():
+    _process_hip_expired_consents()
```
Comment (sravfeyn) [inline abdm_integrator/hip/tasks.py:44]: Why use a dummy method that's not reused and is just one line?

File: abdm_integrator/hiu/tasks.py [hunk 2]
```diff
@@ -14,3 +20,17 @@ def process_hiu_consent_notification_request(self, request_data):
...
+    _process_hiu_expired_consents()
+
+
+def _process_hiu_expired_consents():
+    for consent in ConsentRequest.objects.exclude(status=ConsentStatus.EXPIRED).filter(
+        expiry_date__lt=datetime.utcnow()
+    ):
+        with transaction.atomic():
+            consent.update_status(ConsentStatus.EXPIRED)
+            consent.artefacts.all().delete()
```
Comment (sravfeyn) [inline abdm_integrator/hiu/tasks.py:36]: Is there a way to do this in bulk?

File: abdm_integrator/hip/tasks.py (hunk 1 above)
Comment (ajeety4) [reply abdm_integrator/hip/tasks.py:44]: Only reason for this way as it was easier to write test cases for `_process_hip_expired_consents` without getting into celery related things.

File: abdm_integrator/hiu/tasks.py (hunk 2 above)
Comment (ajeety4) [reply abdm_integrator/hiu/tasks.py:36]: I did think of bulk operation however avoided due to below reasons:
- `artefacts` have cascade relationship with other model and hence cannot use bulk delete for that.
-  Current plan to is to run job every 2 hours which should be frequent enough to have the records count low
(easier to update status in bulk however trickier to delete the related artefacts of consents)
Open for any suggestions for this.

**Review by sravfeyn — APPROVED** [review]

**Pull request merged**